### ISSUE TITLE
Bug: linear part W of layer group symmetry operations

### ISSUE THREAD

**Issue by site-g** [spglib/spglib#377]
I made two mistakes.

- First, I believe that under the Delaunay primitive cell basis, the liner part $\boldsymbol{W}$ of a layer group symmetry operation has the form
```math
\boldsymbol{W} = \begin{pmatrix} W_{11} & W_{12} & W_{13} \\\ W_{21} & W_{22} & W_{23} \\\ 0 & 0 & \pm 1 \end{pmatrix}
```
where $W_{ij}=0,\pm 1$, because I think the non-lattice basis vector $\mathbf{c}'$ after transformation can be the linear combination of old $\mathbf{a}, \mathbf{b}$ and $\mathbf{c}$. However, such a consideration is not compatible with the current workflow of Spglib and will cause error under certain circumstance. Here is an example
```
#include "spglib.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

static void show_layer_spg_dataset(double lattice[3][3],
                                   const double origin_shift[3],
                                   double position[][3],
                                   const int num_atom,
                                   const int types[],
                                   const int aperiodic_axis,
                                   const double symprec);


int main(void)
{
  double lattice[3][3] = {{ 2.0, 0.0,  0.0},
                          { 0.0, 2.0, -1.0},
                          { 0.0, 0.0,  2.0}};
  double origin_shift[3] = {0.0, 0.0, 0.0};
  double position[][3] = {{ 0.0, 0.0, 0.0 }};
  int types[] = {1};
  int num_atom = 1;
  show_layer_spg_dataset(lattice, origin_shift, position, num_atom, types, 2, 1e-5);
}


static void show_layer_spg_dataset(double lattice[3][3],
                                   const double origin_shift[3],
                                   double position[][3],
                                   const int num_atom,
                                   const int types[],
                                   const int aperiodic_axis,
                                   const double symprec)
{
  SpglibDataset *dataset;
  char ptsymbol[6];
  int pt_trans_mat[3][3];

  int i, j, size;
  const char *wl = "abcdefghijklmnopqrstuvwxyz";

  for ( i = 0; i < num_atom; i++ ) {
    for ( j = 0; j < 3; j++ ) {
      position[i][j] += origin_shift[j];
    }
  }

  dataset = spg_get_layer_dataset(lattice,
                                  position,
                                  types,
                                  num_atom,
                                  aperiodic_axis,
                                  symprec);

  printf("International: %s (%d)\n", dataset->international_symbol, dataset->spacegroup_number );
  printf("Hall symbol:   %s\n", dataset->hall_symbol );
  spg_get_pointgroup(ptsymbol,
                     pt_trans_mat,
                     dataset->rotations,
                     dataset->n_operations);
  printf("Point group:   %s\n", ptsymbol);
  printf("Transformation matrix:\n");
  for ( i = 0; i < 3; i++ ) {
    printf("%f %f %f\n",
           dataset->transformation_matrix[i][0],
           dataset->transformation_matrix[i][1],
           dataset->transformation_matrix[i][2]);
  }
  printf("Wyckoff letters:\n");
  for ( i = 0; i < dataset->n_atoms; i++ ) {
    printf("%c ", wl[dataset->wyckoffs[i]]);
  }
  printf("\n");
  printf("Equivalent atoms:\n");
  for (i = 0; i < dataset->n_atoms; i++) {
    printf("%d ", dataset->equivalent_atoms[i]);
  }
  printf("\n");

  for (i = 0; i < dataset->n_operations; i++) {
    printf("--- %d ---\n", i + 1);
    for (j = 0; j < 3; j++) {
      printf("%2d %2d %2d\n",
             dataset->rotations[i][j][0],
             dataset->rotations[i][j][1],
             dataset->rotations[i][j][2]);
    }
    printf("%f %f %f\n",
           dataset->translations[i][0],
           dataset->translations[i][1],
           dataset->translations[i][2]);
  }

  spg_free_dataset(dataset);

}
```
Which is a cell with inclined $\mathbf{c}$ and one atom at origin.
```math
(\mathbf{a}_\mathrm{p}, \mathbf{b}_\mathrm{p}, \mathbf{c}_\mathrm{p}) = \begin{pmatrix} 2 & 0 & 0 \\\ 0 & 2 & -1 \\\ 0 & 0 & 2 \end{pmatrix}
```
The 3D space group of it is 65: _C m m m_ with a cuboid conventional cell
```math
(\mathbf{a}_\mathrm{c}, \mathbf{b}_\mathrm{c}, \mathbf{c}_\mathrm{c}) = \begin{pmatrix} 2 & 0 & 0 \\\ 0 & 4 & 0 \\\ 0 & 0 & 2 \end{pmatrix}
```
Here $`\mathbf{b}_\mathrm{c}`$ is in fact the linear combination of $`\mathbf{b}_\mathrm{p}`$ and $\mathbf{c}_\mathrm{p},$ so is impossible for layer group as it breaks the 2D lattice plane of the layer system. As you can see in the following figure, the norm of the mirror reflection parallel to the $\mathbf{ab}$ plane is $`\mathbf{b}_\mathrm{p} + 2\mathbf{c}_\mathrm{p},`$ where "$`2`$" is illegal.
![image](https://github.com/spglib/spglib/assets/35136455/530b61a5-ca62-4622-b7f8-88731125868d)

In the current implementation of Spglib, neither Delaunay reduction nor symprec can reject this axis choice, and theoretically speaking, Spglib will enter an infinite loop by finding the illegal $`\mathbf{b}_\mathrm{p} + 2\mathbf{c}_\mathrm{p},`$ axis again and again.

- However, such behavior does not appear, because I made another mistake. The candiate $W_{ij}$s are saved in [`relative_axes`](https://github.com/spglib/spglib/blob/8f7f01c5652fbf4cc5bdd0bb02d543f9a1d4a7ab/src/symmetry.c#L54) as column vectors. I mistaken them as row vectors at

https://github.com/spglib/spglib/blob/8f7f01c5652fbf4cc5bdd0bb02d543f9a1d4a7ab/src/symmetry.c#L964-L966
So, in fact what Spglib searches for the linear part of layer group operations now is
```math
\boldsymbol{W} = \begin{pmatrix} W_{11} & W_{12} & 0 \\\ W_{21} & W_{22} & 0 \\\ W_{31} & W_{32} & \pm 1 \end{pmatrix}
```
which rejects $`\mathbf{b}_\mathrm{p} + 2\mathbf{c}_\mathrm{p}.`$

**Comment by site-g** [comment]
So, my question is, how to fix them? Should I make $\boldsymbol{W}$ to be
```math
\boldsymbol{W} = \begin{pmatrix} W_{11} & W_{12} & 0 \\\ W_{21} & W_{22} & 0 \\\ 0 & 0 & \pm 1 \end{pmatrix}
```
This will fix both two errors. But I am concerned that if such fix is reasonable enough.

**Comment by lan496** [comment]
I think we can constrain that the input aperiodic axis is always perpendicular to other periodic axes because I cannot think of any practical cases when the aperiodic axis should be not perpendicular. Then, taking 2D-Delaunay reduction on the periodic axes and searching rotation matrices from 

$$
\begin{pmatrix}
	W_{11} & W_{12} & 0 \\
	W_{21} & W_{22} & 0 \\
	0 & 0 & \pm 1 \\
\end{pmatrix}
$$

may solve this issue 🤔

**Comment by site-g** [comment]
I thought of it. Now I think no matter whether the input aperiodic axis is perpendicular or not, $\boldsymbol{W}$ is always
```math
\begin{pmatrix} W_{11} & W_{12} & 0 \\\ W_{21} & W_{22} & 0 \\\ 0 & 0 & \pm 1 \end{pmatrix} \tag{eq1}
```
Because for layer system, in-plane 3, 4, 6-fold rotations are rejected, so the only special case is $\mathbf{c}_\mathrm{c}$ of the conventional cell being perpendicular to the $\mathbf{a}\mathbf{b}$ plane. This can be achieved in two ways:

1. $`\mathbf{c}_\mathrm{p}`$ of the standard primitive cell is perpendicular to the $`\mathbf{a}\mathbf{b}`$ plane.
2. $`\mathbf{c}_\mathrm{p}`$ of the standard primitive cell points to the face/body center of the conventional cell.

Obviously, the second condition cannot appear for layer group as it assumes translational symmetry along $`\mathbf{c}_\mathrm{p}.`$ So, if any symmtry imposes any constraint on the direction of $\mathbf{c},$ there must exist a primitive cell with perpendicular $`\mathbf{c}_\mathrm{p}.`$

Then we should notice that, if a basis vector of the primitive cell can be transformed to be perpendicular, Delaunay reduction always transforms it. Under this circumstance ([eq1](#mjx-eqn-eq)) always holds. If Delaunay reduction cannot make $`\mathbf{c}_\mathrm{p}`$ perpendicular, and we find a symmetry operation with the form
```math
\begin{pmatrix} W_{11} & W_{12} & W_{13} \\\ W_{21} & W_{22} & W_{23} \\\ 0 & 0 & \pm 1 \end{pmatrix} \tag{eq2}
```
It is in fact the symmetry of a face/body center supercell, and we should reject it.

In conclusion, I think we should use ([eq1](#mjx-eqn-eq)) at any time, because Delaunay reduction will make the aperiodic axis perpendicular when possible.

**Comment by lan496** [comment]
>  if any symmtry imposes any constraint on the direction of c, there must exist a primitive cell with perpendicular c_p.

> In conclusion, I think we should use (eq1) at any time, because Delaunay reduction will make the aperiodic axis perpendicular when possible.

I also think it is correct.

**Comment by atztogo** [comment]
I can not follow you discussion. I have questions.

1. First of all, at least it makes simpler by accepting a convention that a-b axes define the 2D lattice, i.e., fixing c-axis as aperiodic axis. Do you discuss under this condition?

2. What is your Delaunay reduction?

3. Do you distinguish reduction and standardization?

**Comment by site-g** [comment]
1. Yes. Other cases just differ by a rotation.

2. and 3. I mean do
```math
\mathbf{b}_i'=-\mathbf{b}_i,\ \mathbf{b}_j'=-\mathbf{b}_j,\ \mathbf{b}_k'=\mathbf{b}_i+\mathbf{b}_k,\ \mathbf{b}_l'=\mathbf{b}_i+\mathbf{b}_l
```
until all $\mathbf{b}_n \cdot \mathbf{b}_m \le 0\ (1 \le n \le m \le 4)$. Then chose ths shortest 3 vectors among ${\mathbf{b}_1,\mathbf{b}_2,\mathbf{b}_3,\mathbf{b}_4,\mathbf{b}_1+\mathbf{b}_2,\mathbf{b}_2+\mathbf{b}_3,\mathbf{b}_3+\mathbf{b}_1}$ as the basis vectors. Do you mean the first step is Delaunay reduction, and the second step is some kind of standardization? If that is the case, then I did not distinguish reduction and standardization.
